Sources/ProcedureKit/Procedure.swift

```diff
@@ -531,73 +767,171 @@ open class Procedure: Operation, ProcedureProtocol {
+            var queueFinishForStart = false
+            if _state < .started {
+                guard _isCancelled else {
+                    assertionFailure("Cannot finish Procedure prior to it being started (unless cancelled).")
+                    return nil
+                }
+                // if the Procedure is cancelled, we can queue the finish for after
+                // start() is called
+                queueFinishForStart = true
+            }
             // Only a single call to _finish should continue
-            guard !_isHandlingFinish
-                // cancel() and main() ensure one-time execution
-                // thus, cancel() and main() set _isFinishingFrom prior to calling _finish()
-                // but finish() does not; _isFinishingFrom is set below when finish() calls _finish()
-                // this could be simplified to a check for (_isFinishFrom == source) if finish()
-                // ensured that it could only call _finish() once
-                // (although that would require another aquisition of the lock)
-                || (_isFinishingFrom == source && (source == .cancel || source == .main))
-                else { return false }
+            guard !_isHandlingFinish else { return nil }
+            _isHandlingFinish = true
+
+            guard !queueFinishForStart else {
+                // Calls to finish() prior to the Procedure starting are
+                // queued to be executed once the queue has started the Procedure.
+                // (As long as the Procedure is cancelled first.)
+                //
+                // (It's an error for an Operation added to an OperationQueue to
+                // set isFinished to true prior to the queue starting the Operation.)
+                _pendingFinish = FinishingInfo(receivedErrors: receivedErrors, source: source)
+                return nil
+            }
+
+            return FinishingInfo(receivedErrors: receivedErrors, source: source)
+        }
+    }
+
+    private final func finish(withErrors receivedErrors: [Error], from source: ProcedureKit.FinishingFrom) {
+        guard let finishingInfo = shouldFinish(withErrors: receivedErrors, from: source) else {
+            log.verbose(message: "An earlier call to finish \((isFinished) ? "has already succeeded." : "is pending. The Procedure will finish from the first call.") This call will have no effect: finish(withErrors: \(receivedErrors))")
+            return
+        }
+
+        _finish(withInfo: finishingInfo)
+    }
 
-            _isFinishingFrom = source
-            return true
+    private final func _finish(withInfo info: FinishingInfo) {
+        dispatchEvent {
+            self._finishImpl(withInfo: info)
         }
     }
 
-    private final func _finish(withErrors receivedErrors: [Error], from source: ProcedureKit.FinishingFrom) {
-        guard shouldFinish(from: source) else { return }
+    private final func _finishImpl(withInfo info: FinishingInfo) {
```

Comment:
Perhaps the naming here could be a bit more descriptive so it's easier to see why both `_finish` and `_finishImpl` exist. Also `Impl` gives me C++ flashbacks. ;)